subt_ign/src/GameLogicPlugin.cc

```diff
@@ -1865,3 +1874,28 @@ std::ofstream &GameLogicPluginPrivate::Log()
                   << " " << this->simTime.nsec() << " ";
   return this->logStream;
 }
+
+/////////////////////////////////////////////////
+void GameLogicPluginPrivate::CheckRobotFlip(const std::string &_name,
+                                            const ignition::math::Pose3d &_pose)
+{
+  // radian equivalents
+  static const double k170_deg = 2.96706;
+  static const double k190_deg = 3.316126;
+
+  auto orientation = _pose.Rot().Euler();
+  auto roll = std::abs(orientation.X());
```

Comment:
I think Euler angles can be unreliable here because, for example, `(pi, pi, 0) == (0, 0, pi)`, but the vehicle in this case is not flipped. My suggestion would be to multiply the rotation with the unit `z` vector and then take the dot product of the result with the unit `z` vector. The result is the cosine of the angle between the world `+z` axis and the vehicle's `z` axis. If the vehicle is flipped, the value should be close to -1.0.

Reply:
Thanks for pointing that out; I was not aware of that issue with Euler angles.

> My suggestion would be to multiply the rotation with the unit z vector and then take the dot product of the result with the unit z vector. The result is the cosine of the angle between the world +z axis and the vehicle's z axis.

If I am understanding correctly, it sounds like you're suggesting that I use this dot product equation (for two vectors **a** and **b**), isolating `cos(theta)`:

**a** dot **b** = ||**a**|| ||**b**|| cos(theta)   --> cos(theta) = (**a** dot **b**) / (||**a**|| ||**b**||)

So, in this case, **a** would be the rotation, and **b** would be the unit z vector (0,0,1). Am I understanding this correctly?

Reply:
Yeah, the second part of what I was saying would use the dot product equation, but the first part would use the rotation matrix (or quaternion) to first rotate the vehicle's z axis (the unit z vector in the vehicle's local frame) into the world frame. You can use `_pose.Rot()` as a rotation matrix to multiply the unit z vector. So if **R** = `_pose.Rot()` and  **z** is the unit z vector (`0, 0, 1`), then
**a** = **R** * **z**
Now **a** is the vehicle's z axis in the world frame. We can then take the dot product of that with the unit z vector to find if the two are parallel:
cos(theta) = **a** dot **z**. These are all unit vectors so we can ignore magnitudes.

Reply:
That makes sense; thanks for the detailed explanation!

I believe that I have just implemented what you described in [af0b090](https://github.com/osrf/subt/pull/519/commits/af0b0901532ef492eb433d3765f8c569056697ff). Would you mind taking a look at that commit and letting me know if I am still missing something? If it seems correct to you, I can make one last commit that cleans things up, and then I will re-request a review.